### PR TITLE
PLA-22903 - Updated Remediation Job for Key Vault Logging

### PR DESCRIPTION
**Pull request by kshrutik**
Updated the remediation job for Key Vault logging

### CODE REVIEW

**Review by nandeshguru — DISMISSED** [review]

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md [hunk 1]
```diff
@@ -38,13 +41,16 @@ You may run test using following command under vss-remediation-worker-job-code-p
     python3 -m pytest test
 ```
 ## Deployment
-Provision a Virtual Machine Create an EC2 instance to use for the worker. The minimum required specifications are 128 MB memory and 1/2 Core CPU.
-Setup Docker Install Docker on the newly provisioned EC2 instance. You can refer to the [docs here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-basics.html) for more information.
-Deploy the worker image SSH into the EC2 instance and run the command below to deploy the worker image:
+Provision a Virtual Machine Create an Azure Virtual Machine instance to use for the worker. The minimum required specifications are 128 MB memory and 1/2 Core CPU.
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md:44]: Provision an instance by creating an Azure Virtual Machine to use for the worker.

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md [hunk 2]
```diff
@@ -38,13 +41,16 @@ You may run test using following command under vss-remediation-worker-job-code-p
...
+Setup Docker Install Docker on the newly provisioned Azure Virtual Machine instance. You can refer to the [docs here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-basics.html) for more information.
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md:45]: Setup Docker on newly provisioned Azure Virtual Machine instance.

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md [hunk 3]
```diff
@@ -38,13 +41,16 @@ You may run test using following command under vss-remediation-worker-job-code-p
...
+Deploy the worker image SSH into the Azure Virtual Machine instance and run the command below to deploy the worker image:
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/README.md:46]: Deploy the worker docker image by SSH into the Azure Virtual Machine instance and run the following commands.

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 4]
```diff
@@ -16,54 +16,75 @@
...
+    :param subscription_id: Azure Subscription Id
+    :param resource_group_name: Resource group name in which the resource exists
+    :type region: str
+    :type subscription_id: str
+    :type resource_group_name: str
+    :returns: resource name
+    :rtype: str
+    """
+    random_str = "".join(i for i in subscription_id if i.islower() or i.isdigit())
+    subscription_id = random_str[:5]
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:79]: Define 5 as a variable like MAX_COUNT = 5

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 5]
```diff
@@ -16,54 +16,75 @@
...
+    random_str = "".join(i for i in region if i.islower() or i.isdigit())
+    region = random_str[-6:]
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:81]: MAX_COUNT_REGION

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 6]
```diff
@@ -16,54 +16,75 @@
...
+    random_str = "".join(i for i in resource_group_name if i.islower() or i.isdigit())
+    resource_group_name = random_str[-5:]
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:83]: MAX_COUNT_RESOURCE_NAME

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 7]
```diff
@@ -16,54 +16,75 @@
...
-        resource_group_name=resource_group_name,
-        account_name=name,
-        parameters=create_params,
-    )
-    return poller.result()
-
-
 class EnableKeyVaultLogging(object):
     def parse(self, payload):
         """Parse payload received from Remediation Service.
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:90]: Input and output are missing in comment.

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 8]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
 
+    def create_storage_account(
+        self, resource_group_name, name, region, storage_client,
+    ):
+        """Creates a Storage Account
+        :param storage_client: Instance of the Azure StorageManagementClient.
+        :param region: The location in which the storage account exists.
+        :param name: The Storage Account name.
+        :param resource_group_name: The name of the resource group.
+        :type region: str
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:138]: type storage_client str

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 9]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        :type name: str
+        :type resource_group_name: str
+        :return : StorageAccount object
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:141]: type: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 10]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        """
+        from azure.mgmt.storage.models import Sku
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:143]: from azure.mgmt.storage.models importSku as  model_sku

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 11]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        """Updates Storage Account Encryption for a Storage Account.
+        :param storage_client: Instance of the Azure StorageManagementClient.
+        :param resource_group_name: The name of the resource group.
+        :param stg_name: The Storage Account name.
+        :param key_name: Name of the Key to encrypt the Storage Account with.
+        :param vault_uri: Key Vault uri in which the Key exists.
+        :type resource_group_name: str
+        :type stg_name: str
+        :type key_name: str
+        :type vault_uri: str
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:181]: type storage_client: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 12]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        logging.info(f"      key_vault_uri={vault_uri}")
+        logging.info(f"      key_name={key_name}")
+        storage_client.storage_accounts.update(
+            resource_group_name=resource_group_name,
+            account_name=stg_name,
+            parameters=StorageAccountUpdateParameters(
+                encryption=Encryption(
+                    key_source=KeySource.MICROSOFT_KEYVAULT,
+                    key_vault_properties=KeyVaultProperties(
+                        key_name=key_name, key_vault_uri=vault_uri,
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:196]: Take it to new line.

key_name=key_name, 
key_vault_uri=vault_uri,

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 13]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        )
+
+    def check_stg_account(self, storage_client, region, name, resource_group_name):
+        """Checks For the existence of the Storage Account created by CHSS
+        :param storage_client: Instance of the Azure StorageManagementClient.
+        :param region: The location in which the storage account exists.
+        :param name: The Storage Account name.
+        :param resource_group_name: The name of the resource group.
+        :type region: str
+        :type name: str
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:209]: storage_client: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 14]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        :type resource_group_name: str
+        :returns: StorageAccount object
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:211]: type: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 15]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        """
+        storage_accounts_paged: ItemPaged[
+            StorageAccountListResult
+        ] = storage_client.storage_accounts.list()
+        storage_accounts_list: List[dict] = list(storage_accounts_paged)
+        for stg_account in storage_accounts_list:
+            stg_id = stg_account.id
+            stg_components = stg_id.split("/")
+            resource_grp = stg_components[4]
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:220]: Check for length, what if 4th index is not there?

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 16]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+            if (
+                stg_account.name == name
+                and stg_account.location == region
+                and resource_grp == resource_group_name
+            ):
+                return stg_account
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:226]: stroage_account = None at the start

Run the for loop and in the for loop, storage_account = stg_account

return storage_account

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 17]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+
+    def check_key_vault(self, keyvault_client, region, name, resource_group_name):
+        """Checks for the existence of the Key Vault created by CHSS.
+        :param keyvault_client: Instance of the Azure KeyVaultManagementClient.
+        :param region: The location in which the storage account exists.
+        :param name: The Storage Account name.
+        :param resource_group_name: The name of the resource group.
+        :type region: str
+        :type name: str
+        :type resource_group_name: str
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:237]: type keyvault_client: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 18]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        :returns: Vault Object
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:238]: type: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 19]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+        """
+        key_vault_paged: ItemPaged[
+            VaultListResult
+        ] = keyvault_client.vaults.list_by_subscription()
+        key_vault_list: List[dict] = list(key_vault_paged)
+        for key_vault in key_vault_list:
+            key_vault_id = key_vault.id
+            key_vault_components = key_vault_id.split("/")
+            resource_grp = key_vault_components[4]
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:247]: Check for index if we have 5 elements.

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 20]
```diff
@@ -106,16 +127,349 @@ def parse(self, payload):
...
+    ):
+        """Creates a diagnostic setting
+        :param monitor_client: Instance of the Azure StorageManagementClient.
+        :param key_vault_id: The resource Id of the Key Vault.
+        :param key_vault_name: Name of the Key Vault.
+        :param stg_account_id: The Storage Account resource Id.
+        :param log: Instance of Azure Monitor LogSettings
+        :type key_vault_id: str
+        :type key_vault_name: str
+        :type stg_account_id: str
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:267]: param monitor_client: object

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 21]
```diff
@@ -129,42 +483,111 @@ def remediate(
...
-            logging.info("    Creating a Storage Account")
-            logging.info("    executing client_storage.storage_accounts.begin_create")
-            logging.info(f"      resource_group_name={resource_group_name}")
-            logging.info(f"      account_name={stg_name}")
-            stg_account = create_storage_account(
-                resource_group_name, stg_name, region, storage_client
+            key_vault = keyvault_client.vaults.get(
+                resource_group_name=resource_group_name, vault_name=key_vault_name,
             )
+            key_vault_id = key_vault.id
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:490]: Directly use key_vault.id

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 22]
```diff
@@ -129,42 +483,111 @@ def remediate(
...
-            logging.info(f"      resource_uri={key_vault.id}")
-            logging.info(f"      name={key_vault_name}")
-
-            monitor_client.diagnostic_settings.create_or_update(
-                resource_uri=key_vault.id,
-                name=key_vault_name,
-                parameters=DiagnosticSettingsResource(
-                    storage_account_id=stg_account.id, logs=[log],
-                ),
+            app_object_id = app_details.value
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:499]: Directly use app_details.value

File: remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py [hunk 23]
```diff
@@ -129,42 +483,111 @@ def remediate(
...
+                    # Encrypt the Storage Account using the above Key
+                    self.update_storage_account_encryption(
+                        storage_client,
+                        resource_group_name,
+                        stg_account.name,
+                        key.name,
+                        key_vault.properties.vault_uri,
+                    )
+
+                # Creating Diagnostic settings to store violated Key vault's logs
```
Comment (mngaonkar) [inline remediation_worker/jobs/azure_key_vault_logging_for_keyvault_enabled/azure_key_vault_logging_for_keyvault_enabled.py:579]: Else is not required, following part of the code can be out of if:

```
# Creating Diagnostic settings to store violated Key vault's logs
                self.create_diagnostic_setting(
                    monitor_client, key_vault_id, key_vault_name, stg_account.id, log
```

**Review by mngaonkar — DISMISSED** [review]

**Review by mngaonkar — APPROVED** [review]

**Pull request merged**